From 1fc5f0d8398e2c377be154b0b61c8baeb9e17d5f Mon Sep 17 00:00:00 2001
From: Guilin Zhong <guilin.zhong@intel.com>
Date: Thu, 25 Dec 2014 10:48:49 +0800
Subject: [PATCH 198/210] PMIC: Fix a bug in camera sensor powerup

Fix a bug in camera sensor powerup. Pmic won't enable when power supply
enable pin is used for camera module. In most cases both enable pin
and pmic should be enabled when sensor powerup

Change-Id: Ie9dc11cac56e33309ce310aae0517477fe4acb8e
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-15068
Signed-off-by: Guilin Zhong <guilin.zhong@intel.com>
---
 arch/x86/platform/intel-mid/atomisp_gmin_platform.c | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/arch/x86/platform/intel-mid/atomisp_gmin_platform.c b/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
index b569a5e..2e0d611 100644
--- a/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
+++ b/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
@@ -511,10 +511,8 @@ int gmin_v1p8_ctrl(struct v4l2_subdev *subdev, int on)
 		return 0;
 	gs->v1p8_on = on;
 
-	if (v1p8_gpio >= 0) {
+	if (v1p8_gpio >= 0)
 		gpio_set_value(v1p8_gpio, on);
-		return 0;
-	}
 
 	if (gs->v1p8_reg) {
 		if (on)
@@ -538,6 +536,15 @@ int gmin_v1p8_ctrl(struct v4l2_subdev *subdev, int on)
 		return ret;
 	}
 
+	if (pmic_id == PMIC_CRYSTALCOVE) {
+		if (on)
+			return intel_soc_pmic_writeb(CRYSTAL_1P8V_REG,
+								CRYSTAL_ON);
+		else
+			return intel_soc_pmic_writeb(CRYSTAL_1P8V_REG,
+								CRYSTAL_OFF);
+	}
+
 	if (pmic_id == PMIC_WHISKEYCOVE) {
 		int val = intel_soc_pmic_readb(WCOVE_V1P8SX_CTRL);
 		if (on)
@@ -572,10 +579,8 @@ int gmin_v2p8_ctrl(struct v4l2_subdev *subdev, int on)
 		return 0;
 	gs->v2p8_on = on;
 
-	if (v2p8_gpio >= 0) {
+	if (v2p8_gpio >= 0)
 		gpio_set_value(v2p8_gpio, on);
-		return 0;
-	}
 
 	if (gs->v2p8_reg) {
 		if (on)
-- 
1.9.1

